### PR TITLE
OCPBUGS-123512: fix(e2e): derive node runtimes from upgraded NodePool state

### PR DESCRIPTION
**Pull request by mgencur**
## What this PR does / why we need it:

* Use the observed OS image stream and current NodePool version when validating runtime handlers so cross-major upgrades do not retain the pre-upgrade release gate.
* Validate runtime handlers before and after NodePool upgrades and cover RHEL 9, RHEL 10, and legacy fallbacks.
* Validate runtime handlers separately for each node instead of passing the test when the first node passes the check even if other nodes did not.

## Which issue(s) this PR fixes:
Fixes https://redhat.atlassian.net/browse/OCPBUGS-123512

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved NodePool upgrade validation by using current API state before and after upgrades.
  * Runtime validation now accounts for node image streams and version information, including RHEL 9 and RHEL 10 configurations.
  * Improved handling of missing, inconsistent, or invalid image stream and version data.

* **Tests**
  * Added coverage for runtime handler selection across supported node versions and image configurations.
  * Strengthened end-to-end validation of node runtime state during NodePool upgrades.

### CODE REVIEW

File: test/e2e/nodepool_major_version_upgrade_test.go [hunk 1]
```diff
@@ -176,7 +177,9 @@ func (ru *NodePoolMajorVersionUpgradeTest) Run(t *testing.T, nodePool hyperv1.No
 	)
 
 	newNodes := e2eutil.WaitForReadyNodesByNodePool(t, ctx, ru.hostedClusterClient, &nodePool, ru.hostedCluster.Spec.Platform.Type)
-	e2eutil.EnsureNodesRuntime(t, newNodes, &nodePool)
+	upgradedNodePool := &hyperv1.NodePool{}
+	g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), upgradedNodePool)).To(Succeed(), "failed to get NodePool after upgrade")
+	e2eutil.EnsureNodesRuntime(t, newNodes, upgradedNodePool)
```
[the remark this answers is not in the document]
Comment (mgencur) [reply test/e2e/nodepool_major_version_upgrade_test.go]: Fixed

File: test/e2e/nodepool_upgrade_test.go [hunk 2]
```diff
@@ -176,6 +176,9 @@ func (ru *NodePoolUpgradeTest) Run(t *testing.T, nodePool hyperv1.NodePool, node
 
 	t.Logf("Validating all Nodes have the synced labels and taints")
 	e2eutil.EnsureNodesLabelsAndTaints(t, nodePool, nodes)
+	initialNodePool := &hyperv1.NodePool{}
+	g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), initialNodePool)).To(Succeed(), "failed to get NodePool before upgrade")
+	e2eutil.EnsureNodesRuntime(t, nodes, initialNodePool)
```
[the remark this answers is not in the document]
Comment (mgencur) [reply test/e2e/nodepool_upgrade_test.go]: Fixed

File: test/e2e/util/util.go [hunk 3]
```diff
@@ -1382,6 +1374,52 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
+
+	switch nodePool.Spec.OSImageStream.Name {
+	case string(hyperv1.OSImageStreamRHEL9):
+		return true
+	case string(hyperv1.OSImageStreamRHEL10):
+		return false
+	}
+
+	if nodePool.Status.Version != "" {
+		if version, err := semver.ParseTolerant(nodePool.Status.Version); err == nil {
```
[the remark this answers is not in the document]
Comment (mgencur) [reply test/e2e/util/util.go]: Fixed

File: test/e2e/util/util.go [hunk 4]
```diff
@@ -1358,17 +1359,17 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
-		nodePool.Spec.OSImageStream.Name == string(hyperv1.OSImageStreamRHEL9)
-
-	validHandlers := map[string]bool{
-		"crun": false,
-	}
-	if isRHEL9 {
-		validHandlers["runc"] = false
+	expectedHandlers, err := expectedNodeRuntimeHandlers(nodePool)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to determine expected runtime handlers")
+	if err != nil {
```
Comment (jparrill) [inline test/e2e/util/util.go]: This `if err != nil { return }` block is dead code — `g.Expect(err).NotTo(HaveOccurred())` on the previous line calls `t.Fatal` if `err != nil`, so this branch is never reached. I'd drop it to avoid confusion about whether the error could be non-fatal.

File: test/e2e/util/util.go [hunk 5]
```diff
@@ -1358,17 +1359,17 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
+		return
 	}
 
 	for _, node := range nodes {
+		validHandlers := make(map[string]bool, len(expectedHandlers))
```
Comment (jparrill) [inline test/e2e/util/util.go]: Nice catch on a latent bug here — the old code shared one `validHandlers` map across all nodes, so if node 1 already set `"crun": true`, node 2 would inherit that and pass validation even if it was missing the handler. The per-node copy fixes this. Worth mentioning in the PR description since it's a behavior change beyond the stated fix.

File: test/e2e/util/util_test.go [hunk 6]
```diff
@@ -131,6 +133,91 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			},
+			wantRunc: true,
+		},
+		{
+			name:           "legacy suite version is the final fallback",
+			releaseVersion: Version423,
+			nodePool:       &hyperv1.NodePool{},
+			wantRunc:       true,
+		},
+	}
```
Comment (jparrill) [inline test/e2e/util/util_test.go:219]: Two edge cases missing from the test table:
- `nodePool == nil` — `usesRHEL9NodePool` has an explicit nil guard that falls back to `IsLessThan(Version50)`, but nothing exercises it.
- Status vs Spec conflict (e.g. `Status.OSImageStream = rhel-9`, `Spec.OSImageStream = rhel-10`) — the function should return `true` (status wins), but there's no test documenting that invariant.

File: test/e2e/nodepool_upgrade_test.go [hunk 7]
```diff
@@ -190,6 +193,26 @@ func (ru *NodePoolUpgradeTest) Run(t *testing.T, nodePool hyperv1.NodePool, node
...
+				err := ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), np)
+				return np, err
+			},
+			[]e2eutil.Predicate[*hyperv1.NodePool]{
+				e2eutil.OSImageStreamPredicate(expectedStream),
+			},
+			e2eutil.WithTimeout(2*time.Minute),
+		)
+	}
+	initialNodePool := &hyperv1.NodePool{}
```
Comment (jparrill) [inline test/e2e/nodepool_upgrade_test.go]: Minor naming nit: `initialNodePool` is fetched after the `EventuallyObject` wait that polls the API server, so the NodePool may have been reconciled further by this point. `preUpgradeNodePool` would make the temporal position clearer.

**Review by jparrill — COMMENTED** [review]
Nice fix — deriving runtime expectations from actual NodePool state instead of the suite global is the right call. A few minor suggestions inline.

**Review by sdminonne — COMMENTED** [review]
`test/e2e/util/util_test.go` — `TestExpectedNodeRuntimeHandlers`

The test table covers `Status=rhel-9` overriding `Spec=rhel-10` but not the inverse. The inverse is the more critical direction for this PR's scenario: after a 4.x→5.0 upgrade, spec might still say `rhel-9` while status reports `rhel-10` once machines converge. Adding this case would guard against an accidental swap of the two switch blocks in `usesRHEL9NodePool`:

```go
{
    name:           "observed RHEL 10 takes precedence over requested RHEL 9",
    releaseVersion: Version423,
    nodePool: &hyperv1.NodePool{
        Spec:   hyperv1.NodePoolSpec{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9}},
        Status: hyperv1.NodePoolStatus{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL10}},
    },
    wantRunc: false,
},
```

The existing "observed RHEL 9 takes precedence over requested RHEL 10" case only proves status wins when it adds a handler (runc). This case proves status also wins when it removes one — which is the actual upgrade path this PR fixes.

File: test/e2e/util/util_test.go [hunk 8]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+		},
+		{
+			name:           "observed RHEL 9 takes precedence over requested RHEL 10",
+			releaseVersion: Version50,
+			nodePool: &hyperv1.NodePool{
+				Spec:   hyperv1.NodePoolSpec{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL10}},
+				Status: hyperv1.NodePoolStatus{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9}},
+			},
+			wantRunc: true,
+		},
```
Comment (sdminonne) [inline test/e2e/util/util_test.go:180]: Consider adding the inverse precedence test case here. The existing case proves status wins when it *adds* a handler (RHEL 9 adds runc). The inverse proves status also wins when it *removes* one — which is the actual 4.x→5.0 upgrade path this PR fixes. It would also guard against an accidental swap of the two switch blocks in `usesRHEL9NodePool`.

```suggestion
		},
		{
			name:           "observed RHEL 10 takes precedence over requested RHEL 9",
			releaseVersion: Version423,
			nodePool: &hyperv1.NodePool{
				Spec:   hyperv1.NodePoolSpec{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9}},
				Status: hyperv1.NodePoolStatus{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL10}},
			},
			wantRunc: false,
```

File: test/e2e/util/util_test.go [hunk 9]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+
+	tests := []struct {
+		name           string
+		releaseVersion semver.Version
+		nodePool       *hyperv1.NodePool
+		wantRunc       bool
+		wantErr        bool
+	}{
+		{
+			name:           "observed RHEL 10 overrides an older suite release version",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]: nit: Per [TESTING.md](https://github.com/openshift/hypershift/blob/main/TESTING.md), test case names must follow the `"When <condition>, it should <expected behavior>"` format. All 9 test cases here use descriptive phrases instead.

For example:
```
// Current
"observed RHEL 10 overrides an older suite release version"
"observed RHEL 9 takes precedence over requested RHEL 10"
"status version falls back to RHEL 9"

// Convention
"When status reports RHEL 10 and suite is pre-5.0, it should not require runc"
"When status reports RHEL 9 and spec requests RHEL 10, it should require runc"
"When status version is pre-5.0 with no stream info, it should require runc"
```

File: test/e2e/nodepool_major_version_upgrade_test.go [hunk 10]
```diff
@@ -123,6 +123,7 @@ func (ru *NodePoolMajorVersionUpgradeTest) Run(t *testing.T, nodePool hyperv1.No
 		np := &hyperv1.NodePool{}
 		g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), np)).To(Succeed())
 		t.Logf("Pre-upgrade osImageStream: %q", np.Status.OSImageStream.Name)
+		e2eutil.EnsureNodesRuntime(t, nodes, np)
```
Comment (sdminonne) [inline test/e2e/nodepool_major_version_upgrade_test.go]: This `EnsureNodesRuntime` call is placed inside the bare block scoped by the comment `// Record pre-upgrade osImageStream.` (line 121), which was originally just logging. The block now does two things (record + validate) but the comment only describes one.

For consistency with `nodepool_upgrade_test.go` (which fetches a dedicated `preUpgradeNodePool` outside any block), consider moving this out of the block:

```go
	// Record pre-upgrade osImageStream.
	{
		np := &hyperv1.NodePool{}
		g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), np)).To(Succeed())
		t.Logf("Pre-upgrade osImageStream: %q", np.Status.OSImageStream.Name)
	}

	preUpgradeNodePool := &hyperv1.NodePool{}
	g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), preUpgradeNodePool)).To(Succeed(), "failed to get NodePool before upgrade")
	e2eutil.EnsureNodesRuntime(t, nodes, preUpgradeNodePool)
```

Alternatively, just update the comment to `// Record pre-upgrade osImageStream and validate runtime handlers.` and reuse the existing `np`.

File: test/e2e/util/util_test.go (hunk 9 above)
Comment (sdminonne) [inline test/e2e/util/util_test.go]: nit: Per [TESTING.md](https://github.com/openshift/hypershift/blob/main/TESTING.md), test case names must follow `"When <condition>, it should <expected behavior>"`. Applies to all 9 cases.

```suggestion
			name:           "When status reports RHEL 10 and suite is pre-5.0, it should not require runc",
```

File: test/e2e/util/util_test.go [hunk 11]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version423,
+			nodePool: &hyperv1.NodePool{Status: hyperv1.NodePoolStatus{
+				OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL10},
+			}},
+			wantRunc: false,
+		},
+		{
+			name:           "observed RHEL 9 overrides a newer suite release version",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When status reports RHEL 9 and suite is 5.0+, it should require runc",
```

File: test/e2e/util/util_test.go [hunk 12]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version50,
+			nodePool: &hyperv1.NodePool{Status: hyperv1.NodePoolStatus{
+				OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9},
+			}},
+			wantRunc: true,
+		},
+		{
+			name:           "observed RHEL 9 takes precedence over requested RHEL 10",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When status reports RHEL 9 and spec requests RHEL 10, it should require runc",
```

File: test/e2e/util/util_test.go [hunk 13]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version50,
+			nodePool: &hyperv1.NodePool{
+				Spec:   hyperv1.NodePoolSpec{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL10}},
+				Status: hyperv1.NodePoolStatus{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9}},
+			},
+			wantRunc: true,
+		},
+		{
+			name:           "status version falls back to RHEL 9",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When status version is pre-5.0 with no stream info, it should require runc",
```

File: test/e2e/util/util_test.go [hunk 14]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version50,
+			nodePool:       &hyperv1.NodePool{Status: hyperv1.NodePoolStatus{Version: "4.23.0"}},
+			wantRunc:       true,
+		},
+		{
+			name:           "status version falls back to RHEL 10",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When status version is 5.0+ with no stream info, it should not require runc",
```

File: test/e2e/util/util_test.go [hunk 15]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version423,
+			nodePool:       &hyperv1.NodePool{Status: hyperv1.NodePoolStatus{Version: "5.0.0"}},
+			wantRunc:       false,
+		},
+		{
+			name:           "invalid status version fails validation",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When status version is invalid semver, it should return an error",
```

File: test/e2e/util/util_test.go [hunk 16]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version50,
+			nodePool:       &hyperv1.NodePool{Status: hyperv1.NodePoolStatus{Version: "not-a-semver"}},
+			wantErr:        true,
+		},
+		{
+			name:           "explicit RHEL 9 stream overrides status version",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When spec requests RHEL 9 and status version is 5.0+, it should require runc",
```

File: test/e2e/util/util_test.go [hunk 17]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version50,
+			nodePool: &hyperv1.NodePool{
+				Spec:   hyperv1.NodePoolSpec{OSImageStream: hyperv1.OSImageStreamReference{Name: hyperv1.OSImageStreamRHEL9}},
+				Status: hyperv1.NodePoolStatus{Version: "5.0.0"},
+			},
+			wantRunc: true,
+		},
+		{
+			name:           "legacy suite version is the final fallback",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When NodePool has no stream or version info, it should fall back to suite release version",
```

File: test/e2e/util/util_test.go [hunk 18]
```diff
@@ -131,6 +133,105 @@ func TestAllowedCIDRsTargetService(t *testing.T) {
...
+			releaseVersion: Version423,
+			nodePool:       &hyperv1.NodePool{},
+			wantRunc:       true,
+		},
+		{
+			name:           "nil NodePool uses the legacy suite version fallback",
```
Comment (sdminonne) [inline test/e2e/util/util_test.go]:
```suggestion
			name:           "When NodePool is nil, it should fall back to suite release version",
```

File: test/e2e/nodepool_major_version_upgrade_test.go [hunk 19]
```diff
@@ -123,6 +123,7 @@ func (ru *NodePoolMajorVersionUpgradeTest) Run(t *testing.T, nodePool hyperv1.No
...
 	}
 
 	// Upgrade to latest release.
```
Comment (sdminonne) [inline test/e2e/nodepool_major_version_upgrade_test.go]: The `EnsureNodesRuntime` call is inside a block scoped by `// Record pre-upgrade osImageStream.` (line 121), which was originally just logging. Move it outside the block for consistency with `nodepool_upgrade_test.go`'s `preUpgradeNodePool` pattern.

```suggestion
	}

	preUpgradeNodePool := &hyperv1.NodePool{}
	g.Expect(ru.mgmtClient.Get(ctx, crclient.ObjectKeyFromObject(&nodePool), preUpgradeNodePool)).To(Succeed(), "failed to get NodePool before upgrade")
	e2eutil.EnsureNodesRuntime(t, nodes, preUpgradeNodePool)

	// Upgrade to latest release.
```

File: test/e2e/util/util.go [hunk 20]
```diff
@@ -1382,6 +1380,58 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
+	if err != nil {
+		return nil, err
+	}
+	if usesRHEL9 {
+		validHandlers["runc"] = false
+	}
+	return validHandlers, nil
+}
+
+func usesRHEL9NodePool(nodePool *hyperv1.NodePool) (bool, error) {
```
Comment (sdminonne) [inline test/e2e/util/util.go:1402]: nit: Missing doc comment. Its sibling `expectedNodeRuntimeHandlers` has one.

```suggestion
// usesRHEL9NodePool reports whether the given NodePool runs RHEL 9 nodes.
func usesRHEL9NodePool(nodePool *hyperv1.NodePool) (bool, error) {
```

File: test/e2e/util/util.go [hunk 21]
```diff
@@ -1382,6 +1380,58 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
+	if nodePool == nil {
+		return IsLessThan(Version50), nil
+	}
+
+	// Status reports the stream observed on the nodes and therefore takes precedence
+	// over the requested stream during and after a rollout.
+	switch nodePool.Status.OSImageStream.Name {
+	case string(hyperv1.OSImageStreamRHEL9):
+		return true, nil
+	case string(hyperv1.OSImageStreamRHEL10):
```
Comment (sdminonne) [inline test/e2e/util/util.go]: nit: `OSImageStreamRHEL9` and `OSImageStreamRHEL10` are untyped string constants and `Name` is `string` — the `string()` casts are no-ops. Same applies to the Spec switch below (lines 1418, 1420).

```suggestion
	case hyperv1.OSImageStreamRHEL9:
		return true, nil
	case hyperv1.OSImageStreamRHEL10:
```

File: test/e2e/util/util.go [hunk 22]
```diff
@@ -1382,6 +1380,58 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
+		return false, nil
+	}
+
+	switch nodePool.Spec.OSImageStream.Name {
+	case string(hyperv1.OSImageStreamRHEL9):
+		return true, nil
+	case string(hyperv1.OSImageStreamRHEL10):
```
Comment (sdminonne) [inline test/e2e/util/util.go]:
```suggestion
	case hyperv1.OSImageStreamRHEL9:
		return true, nil
	case hyperv1.OSImageStreamRHEL10:
```

File: test/e2e/util/util.go [hunk 23]
```diff
@@ -1358,17 +1359,14 @@ func EnsureNodesRuntime(t *testing.T, nodes []corev1.Node, nodePool *hyperv1.Nod
...
-		validHandlers["runc"] = false
-	}
+	expectedHandlers, err := expectedNodeRuntimeHandlers(nodePool)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to determine expected runtime handlers")
 
 	for _, node := range nodes {
+		validHandlers := make(map[string]bool, len(expectedHandlers))
+		for handler := range expectedHandlers {
+			validHandlers[handler] = false
+		}
```
Comment (sdminonne) [inline test/e2e/util/util.go]: nit: `maps.Clone` (standard library since Go 1.21) replaces this manual copy. Values in `expectedHandlers` are already `false`, so the clone is semantically identical. Requires adding `"maps"` to the import block.

```suggestion
		validHandlers := maps.Clone(expectedHandlers)
```

**Review by sdminonne — COMMENTED** [review]
nit: Test naming convention and minor code cleanup suggestions. All comments include ```suggestion``` blocks for easy application.

**Review by sdminonne — APPROVED** [review]
LGTM — all review feedback addressed.
